scheduler/test/cook/test/simulator.clj

```diff
@@ -0,0 +1,565 @@
+    ;; We are setting time to enable us to have deterministic runs
+    ;; of the simulator while hooking into the scheduler as non-invasively
+    ;; as possible. A longer explanation can be found in the simulator dev docs.
+    (org.joda.time.DateTimeUtils/setCurrentMillisFixed initial-time)
+    (log/info "Starting simulation.")
+    (with-cook-scheduler mesos-datomic-conn make-mesos-driver-fn
+      scheduler-config
+      (try
+        (doseq [{:keys [user mem cpus gpus]} (:shares config)]
+          (share/set-share! mesos-datomic-conn user "simulation" :mem mem :cpus cpus :gpus gpus))
+        (loop [trace trace
+               simulation-time simulation-time
+               fake-real-time initial-time]
+
+          (org.joda.time.DateTimeUtils/setCurrentMillisFixed fake-real-time)
+
+          (let [start-ms (System/currentTimeMillis)
+                submission-batch (take-while #(<= (:submit-time-ms %) simulation-time) trace)
+                rank-complete-chan (async/chan)
+                rebalancer-complete-chan (async/chan)]
+            (log/info "Simulation time: " simulation-time)
+            (log/info "submission-batch size: " (count submission-batch))
+
+            (log/info "Submitting batch")
+            ;; Submit new jobs
+            (doseq [job submission-batch]
+              (org.joda.time.DateTimeUtils/setCurrentMillisFixed (inc (.getTime (tc/to-date (t/now)))))
+              (submit-job mesos-datomic-conn job))
+            ;; Ensure peer has acknowledged the new jobs
+            (when (seq submission-batch)
+              (poll-until #(d/q '[:find ?j .
+                                      :in $ ?t
+                                      :where
+                                      [?j :job/submit-time ?t]]
+                                    (d/db mesos-datomic-conn)
+                                    ;; This relies on
+                                    ;; 1. Time is controlled
+                                    ;; 2. We increment time for each job submitted
+                                    (tc/to-date (t/now)))
+                        50
+                        60000))
+            (log/info "Batch submission complete")
+
+            ;; Rebalance
+            (log/info "Starting rebalance")
+            (org.joda.time.DateTimeUtils/setCurrentMillisFixed (inc (.getTime (tc/to-date (t/now)))))
+            (async/>!! rebalancer-trigger-chan rebalancer-complete-chan)
+            (async/<!! rebalancer-complete-chan)
+            (log/info "Rebalance complete")
+
+            ;; Request for jobs that are complete to have cook be notified
+            (log/info "Send completion status to scheduler")
+            (org.joda.time.DateTimeUtils/setCurrentMillisFixed (inc (.getTime (tc/to-date (t/now)))))
+            (async/>!! complete-trigger-chan :trigger)
+            (async/>!! complete-trigger-chan :trigger)
```

Comment:
Please add a comment that we are relying on the processing of this second trigger to be a no-op.